vm_insnhelper.c

```diff
@@ -2554,30 +2554,81 @@ vm_call_sorbet_simple_p(const rb_method_sorbet_t *sorbet)
+        param->flags.has_kw == TRUE &&
+        param->flags.has_kwrest == FALSE &&
+        param->flags.has_block == FALSE;
+}
+
+enum sorbet_method_opt_kind {
+    /* Not specially optimizable */
+    SORBET_METHOD_OPT_NONE,
+    /* Only required parameters */
+    SORBET_METHOD_OPT_REQ_PARAM_ONLY,
+    /* Keyword arguments without potential kwsplats */
+    SORBET_METHOD_OPT_EFFICIENT_KWARGS,
+};
+
+/* This call combines vm_call_iseq_optimizable_p and logic in vm_callee_setup_arg */
+static enum sorbet_method_opt_kind
 vm_call_sorbet_optimizable_p(const struct rb_call_info *ci, const struct rb_call_cache *cc,
-                             const rb_method_sorbet_t *sorbet)
+                             const struct rb_calling_info *calling, const rb_method_sorbet_t *sorbet)
 {
-    if (!vm_call_iseq_optimizable_p(ci, cc)) {
-        return false;
+    /* Splat calls are generally not interesting, because they can introduce kwsplats
+     * and require extra processing anyway to resolve the splat. */
+    if (IS_ARGS_SPLAT(ci)) {
+        return SORBET_METHOD_OPT_NONE;
     }
 
-    /* vm_callee_setup_arg */
-    if (UNLIKELY(ci->flag & VM_CALL_KW_SPLAT)) {
-        return false;
+    /* Similarly for keyword splats. */
+    if (IS_ARGS_KW_SPLAT(ci)) {
+        return SORBET_METHOD_OPT_NONE;
     }
 
-    /* We only handle simple calls to functions with positional args, unlike
-     * vm_callee_setup_arg */
-    if (!vm_call_sorbet_simple_p(sorbet)) {
-        return false;
+    /* Protected methods are weird, so don't deal with them. */
+    if (METHOD_ENTRY_VISI(cc->me) == METHOD_VISI_PROTECTED) {
+        return SORBET_METHOD_OPT_NONE;
     }
 
-    /* This callsite is to a method that only takes positional arguments. */
-    /* TODO: change this to handle more of the cases that vm_callee_setup_arg does,
-     * like kwarg-only functions.  */
-    return true;
+    /* Positional-argument-only methods are easy to optimize and very common.  But
+     * make sure we're not providing keyword args here.  */
+    if (vm_call_sorbet_simple_p(sorbet) && !IS_ARGS_KEYWORD(ci)) {
+        return SORBET_METHOD_OPT_REQ_PARAM_ONLY;
+    }
+
+    /* If we're calling a keyword-arg taking function that doesn't have other complex
+     * arguments, we can avoid turning the keyword args into a keyword splat. */
+    if (vm_call_kwarg_simple_p(sorbet) && IS_ARGS_KEYWORD(ci)) {
+        /* Because Ruby keyword args can be rolled up into keyword splats to satisfy
+         * the last positional arg, the only case we can handle is when there are
+         * exactly the right number of positional args already.
+         *
+         * vm_callee_setup_arg has another case for no keyword args getting passed,
+         * but we don't handle that yet (and if we did handle it, we'd do it in the
+         * compiled code instead of in the VM, barring major changes to how the
+         * compiled code's argument parsing works).
+         */
```

Comment:
From what I remember keyword splats and efficient keyword args are mutually exclusive. Given that `IS_ARGS_KEYWORD` is testing for the `VM_CALL_KWARG` flag, do you need to consider the case where the last argument could be a keyword hash?

Maybe I'm misreading this comment, and it's really suggesting that `IS_ARGS_KEYWORD` is necessary to guard further checks that it would be possible to enable the efficient kwargs on this call.

Reply:
That's a good question, but I'm having trouble imagining what such a case would look like or how you would wind up with that.  For cases like:

```ruby
def f(name, arg: :default)
  p name, arg
end

f(args: :passed, as: :hash, arg: :passed_normally)
```

the interpreter will roll the keyword args up into a hash to satisfy the first argument, even though `arg:` was specified as an argument.  And if you pass an explicit hash as the first argument, then you'll have the keywords passed as normal because you have an explicit first argument...though you'll have an error because you passed keyword args the function wasn't expecting.

(The `IS_ARGS_KEYWORD` check is, I think, really meant to be testing "yes, we really are passing keyword arguments to this function", because the no-keywords case is something entirely different.)

Do you have an example of what you're thinking of?

Reply:
OK, I wasn't thinking about the case where not filling out enough positional args would cause the keyword args to be turned into a hash :(

I think that the checks here make sense given that, thanks!